web/src/compose_ui.js

```diff
@@ -682,6 +682,89 @@ export function format_text($textarea, type, inserted_content) {
         );
     };
 
+    // Links have to be formatted differently because formatting is not only
+    // at the beginning and end of the text, but also in the middle
+    // Therefore more checks are necessary if selected text is already formatted
+    const format_link = () => {
+        const link_syntax_start = "[";
+        const link_syntax_end = "](url)";
+
+        // Captures:
+        // [<description>](<url>)
+        // with just <url> selected
+        const is_selection_url = () =>
+            range.start >= "[](".length &&
+            text.length - range.end >= ")".length &&
+            text.slice(range.start - 2, range.start) === "](" &&
+            text[range.end] === ")" &&
+            text.lastIndexOf("[", range.start - 3) < text.lastIndexOf("]", range.start - 2);
+
+        // Captures:
+        // [<description>](<url>)
+        // with just <description> selected
+        const is_selection_description_of_link = () =>
+            range.start >= "[".length &&
+            text.length - range.end >= "]()".length &&
+            text.slice(range.start - 1, range.start) === "[" &&
+            text.slice(range.end, range.end + 2) === "](" &&
+            text.includes(")", range.end + 2) &&
+            (text.includes("(", range.end + 2)
+                ? text.indexOf(")", range.end + 2) < text.indexOf("(", range.end + 2)
+                : true);
+
+        // Captures:
+        // [<description>](<url>)
+        // with [<description>](<url>) selected
+        const is_selection_link = () =>
+            range.length >= "[]()".length &&
+            text[range.start] === "[" &&
+            text[range.end - 1] === ")" &&
+            text.slice(range.start + 1, range.start - 1).includes("](");
+
+        if (is_selection_url()) {
+            const beginning = text.lastIndexOf("[", range.start);
+            const url = selected_text === "url" ? "" : " " + selected_text;
+            text =
+                text.slice(0, beginning) +
+                text.slice(beginning + 1, text.indexOf("]", beginning)) +
+                url +
+                text.slice(range.end + 1);
+            set(field, text);
+            field.setSelectionRange(range.start - 2, range.start - 3 + url.length);
+            return;
+        } else if (is_selection_description_of_link()) {
+            let url = text.slice(range.end + 2, text.indexOf(")", range.end));
+            url = url === "url" ? "" : " " + url;
+            text =
+                text.slice(0, range.start - 1) +
+                text.slice(range.start, range.end) +
+                url +
+                text.slice(text.indexOf(")", range.end) + 1);
```

Comment:
I'm skeptical that `text.indexOf(")"` is going to be what you intend it to be if the message contains a bunch of other Markdown links both before and after the selected description.